pkg/controllers/thanos-querier/controller.go

```diff
@@ -0,0 +1,353 @@
+	}
+
+	var requests []reconcile.Request
+	for _, item := range queriers.Items {
+
+		sel, err := metav1.LabelSelectorAsSelector(&item.Spec.Selector)
+		if err != nil {
+			return []reconcile.Request{}
+		}
+		if sel.Matches(labels.Set(ms.GetLabels())) {
+			logger.Info("Found querier, scheduling sync")
+			requests = append(requests, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      item.GetName(),
+					Namespace: item.GetNamespace(),
+				},
+			})
+		}
+	}
+	return requests
+}
+
+// Filter a list of OwnerReferences and return the first owner of Kind
+// MonitoringStack.
+func findMonStackServiceOwner(owners []metav1.OwnerReference) (error, string) {
+	for _, owner := range owners {
+		// This should be compared to a MonitoringStack{} object, but in
+		// this case Kind is empty, so use hardcoded string for now
+		if owner.Kind == "MonitoringStack" {
+			return nil, owner.Name
+		}
+	}
+	return errors.New("Owner is not of kind MonitoringStack"), ""
+}
+
+// Create an object after setting the owner reference to the passed
+// ThanosQueriers.
+// Uses server-side-apply.
+func (r reconciler) reconcileComponent(ctx context.Context, thanos *msoapi.ThanosQuerier, component client.Object) error {
+	if thanos.Namespace == component.GetNamespace() {
+		if err := controllerutil.SetControllerReference(thanos, component, r.scheme); err != nil {
+			return err
+		}
+	}
+
+	if err := r.Patch(ctx, component, client.Apply, client.ForceOwnership, client.FieldOwner("thanos-querier-controller")); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete all resources attached to the passed ThanosQuerier, if they still
+// exist.
+func (r reconciler) maybeDeleteResources(ctx context.Context, thanos *msoapi.ThanosQuerier) {
+	logger := r.logger.WithValues("thanosQuerier", thanos.GetNamespace()+"/"+thanos.GetName())
+	logger.Info("Deleting resources created by")
+	for _, gvk := range []schema.GroupVersionKind{
+		{
+			Group:   "apps",
+			Kind:    "Deployment",
+			Version: "v1",
+		},
+		{
+			Group:   "",
+			Kind:    "ServiceAccount",
+			Version: "v1",
+		},
+		{
+			Group:   "monitoring.coreos.com",
+			Kind:    "ServiceMonitor",
+			Version: "v1",
+		},
+	} {
+		resource := &unstructured.Unstructured{}
+		resource.SetGroupVersionKind(gvk)
+		if err := r.DeleteAllOf(ctx, resource, client.InNamespace(thanos.Namespace), client.MatchingLabels{"app.kubernetes.io/part-of": "thanos-querier-" + thanos.Name}); err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Info("Resources not found, nothing to do")
+			} else {
+				logger.Error(err, "DeleteAllOf failed", "resource", resource)
+			}
+		}
+	}
+	srv := &corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Service",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "thanos-querier-" + thanos.Name,
+			Namespace: thanos.Namespace,
+		},
+	}
+	if err := r.Delete(ctx, srv); err != nil {
```

Comment:
I think we can ignore this error. We do that in other controllers as well.

Reply:
I left the check for errors other then `IsNotFound` here since we want to delete resources in some cases even when the `ThanosQuerier` object still exists (but the targeted `Stack` object was deleted). I think we want to know about a delete error in this case?

Reply:
Ok that makes sense, we can log the error for debugging purposes.